worker/src/lib.rs

```diff
@@ -0,0 +1,615 @@
+    {
+        unsafe {
+            let response_data = mem::ManuallyDrop::new(response_data);
+            let size = mem::size_of_val(&response_data) as u32;
+            let ptr = &response_data as *const _ as *const c_void;
+            let response_function = if let Some(response_function) = self.response_function {
+                response_function
+            } else {
+                return Err(WorkerError::Unknown);
+            };
+            match (response_function)(self.respond_handle, size, ptr) {
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_SUCCESS => Ok(()),
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_ERR_UNKNOWN => Err(WorkerError::Unknown),
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_ERR_NO_SPACE => Err(WorkerError::NoSpace),
+                _ => Err(WorkerError::Unknown),
+            }
+        }
+    }
+}
+
+/// Errors potentially generated by [`Worker`](trait.Worker.html) methods
+pub enum WorkerError {
+    /// Unknown or general error
+    Unknown,
+    /// Failure due to a lack of space
+    NoSpace,
+}
+
+/// The trait providing worker extension to rust-lv2 plugins.
+///
+/// This trait, and `struct` that implements it, provide methods called by the host to implement
+/// the worker extension.
+///
+/// However, the host will not directly use these methods. Instead, the host use wrapping functions
+/// associated with [`WorkerDescriptor`](struct.WorkerDescriptor.html) and then, you pass this
+/// functions to the host with `match_extensions!` macro in `extension_data()` methods of the
+/// Plugin trait.
+pub trait Worker: Plugin {
+    /// Type of data sent to `work` by using a schedule handler.
+    type WorkData: 'static + Send;
+    /// Type of data sent to `work_response` by using a response handler.
+    type ResponseData: 'static + Send;
+    /// The work to do in a non-real-time context,
+    ///
+    /// This is called by the host in a non-realtime context as requested, probably in a separate
+    /// thread from `run()` and possibly with an arbitrary message to handle.
+    ///
+    /// A response can be sent to `run()` context using the response handler. The plugin MUST NOT make any assumptions
+    /// about which thread calls this method, except that there are no real-time requirements and
+    /// only one call may be executed at a time. That is, the host MAY call this method from any
+    /// non-real-time thread, but MUST NOT make concurrent calls to this method from several
+    /// threads.
+    fn work(
+        &mut self,
+        response_handler: &ResponseHandler<Self>,
+        data: Self::WorkData,
+    ) -> Result<(), WorkerError>;
+
+    /// Handle a response from the worker.
+    ///
+    /// This is called by the host in the `run()` context when a response from the worker is ready.
+    #[allow(unused_variables)]
+    fn work_response(&mut self, data: Self::ResponseData) -> Result<(), WorkerError> {
+        Ok(())
+    }
```

Comment:
I used #[allow(unused_variables)] for aesthetic reason, i found it was ugly to have "_data" in the doc.

for writing output here, i think it's doesn't make a sens since this may not called in the cycle corresponding to the schedule work.

Reply:
Edit: for the output reading/writing here, i checked the code and it's trivial to add ports as parameter.I'll try to check if this can result to an issue, and if not, i will enable it. Same for `end_run`